instrumentation/message-handler/aws/src/main/java/io/opentelemetry/instrumentation/messagehandler/SqsBatchMessageHandler.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.messagehandler;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanLinksExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanNameExtractor;
+import io.opentelemetry.instrumentation.awssdk.v2_2.SqsMessageAccess;
+import io.opentelemetry.instrumentation.awssdk.v2_2.SqsParentContext;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.Collection;
+import java.util.Map;
+import javax.annotation.Nullable;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.services.sqs.model.Message;
+
+public abstract class SqsBatchMessageHandler extends BatchMessageHandler<Message> {
+  public SqsBatchMessageHandler(
+      OpenTelemetry openTelemetry, SpanNameExtractor<Collection<Message>> spanNameExtractor) {
+    super(openTelemetry, spanNameExtractor);
+  }
+
+  public SqsBatchMessageHandler(
+      OpenTelemetry openTelemetry,
+      SpanNameExtractor<Collection<Message>> spanNameExtractor,
+      String messageOperation) {
+    super(openTelemetry, spanNameExtractor, messageOperation);
+  }
+
+  @Override
+  protected void setup() {
+    messageInstrumenter =
+        Instrumenter.<Collection<Message>, Void>builder(
+                getOpenTelemetry(), "io.opentelemetry.message.handler", getSpanNameExtractor())
+            .setInstrumentationVersion("1.0")
+            .addAttributesExtractor(getGenericAttributesExtractor())
+            .addAttributesExtractor(getAttributesExtractor())
+            .addSpanLinksExtractor(getSpanLinksExtractor())
+            .buildInstrumenter();
```

Comment:
By default `Instrumenter` creates internal spans, is this intentional? You don't seem to be following the messaging semantic conventions, is this intentional?

Reply:
Yes, these are not client or server spans.

Reply:
Thanks for the input. I thought it was supposed to be internal.
I have adapted the SqsMessageHandler to use different span kinds depending on the messaging operation.
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving
Defines a scenario where there is a single `receive` span for the message batch that has a child `process` span for each individual message. These `process` spans link back to the `producer`.
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-processing
Defines a scenario where messages are received individually and processed as a batch. Here process span links back to all the producers.
Is your intention to implement one of these 2 scenarios or something else?
Note that the specification gives guidelines for messaging span name https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-name Also note that we have some helper classes for implementing messaging instrumentations in https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/messaging For example if you implement the `MessagingAttributesGetter` you can use `MessagingSpanNameExtractor` to produce a spec compliant span name. Do consider whether these could be of use to you.

Reply:
The original idea for this is a single span is created for a batch of upstream messages that are received.
The class will create the span links and add attributes as needed.
It was expanded to cover processing message nodes as well.
It fit more scenario 1, but we only do one step at a time: receive or process.

Reply:
We generally require the instrumentations to follow the specification if there is one. Please review your code an the specification and try to ensure that the telemetry you produce conforms to the specification. Preferably use the messages helper classes that we have. Doing it will give this code a much better chance to get accepted. 
If you feel that specification does not cover you use case consider submitting a pr against specification or raising an issue in the specification repository or joining the messaging SIG and asking for guidance there. You can of course also elaborate here, or in the comments of your code, why you chose to deviate from the specification, or present your case in the java SIG meeting.
In your place I would at first concentrate on the aws part of this pr. Get the shape of the telemetry to conform to the spec, figure out how it would interact with existing aws sdk instrumentation etc. After you have figured out the aws part it should be simpler to see what kind of abstractions the reusable message handler has to provide or whether building it is feasible at all.

Reply:
I went through the spec and everything looked good. Let me know if there is anything else.

Reply:
Thanks for the heads up.

When the spec is adopted and the semantic conventions are updated, then we will need to update this piece.

Until then, we are in spec as far as I can tell.

Reply:
Currently you support `receive`, `process` and `send` (in spec this has been renamed to `publish`) as messaging operation. The first question is does `send` actually make sense here? Secondly for `receive` the language in spec is

> The receive span is be used to track the time used for receiving the message(s), whereas the process span(s) track the time for processing the message(s).

But in this api you already have received the messages, does allowing user to set operation to `receive` make sense? How do you envision users get the messages they are supposed to pass to your handler? For `process` spec has scenarios for processing messages as single batch or one by one. I'm curious why you chose to process them as a batch instead of processing them one by one? Having one process span seems like a valid option, but processing the messages one by one could perhaps allow extracting more telemetry? Was this something you considered?

Reply:
I see.

I will refactor it to focus on receiving messages. I don't think it makes sense to try to combine it with the processor logic given this.

Reply:
@laurit @Oberon00 
I'm looking at the spec.
We need to add span links when creating the span.
But we won't have the messages to extract the span links from until after the span is created.
This appears to be a catch 22.

How do we go about resolving that?

Should we retroactively change the span creation time? Make it so span links can be added after creation? Or maybe make a child span and add span links to that?

Reply:
> I will refactor it to focus on receiving messages.

IIUC, receive spans should track the receipt, that is these are the spans that are already created by the SDK instrumentation for calls like [receiveMessage](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/sqs/SqsClient.html#receiveMessage(software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest)) that are already created by the SDK instrumentation.

Reply:
I got around this by spoofing the span creation time. Let me know if you see concerns here.

Reply:
This has been changed so that it only covers processing spans.